src/nvim/ex_docmd.c

```diff
@@ -1220,6 +1220,302 @@ static char_u *skip_colon_white(const char_u *p, bool skipleadingwhite)
   return (char_u *)p;
 }
 
+/// Set the addr type for command
+///
+/// @param p pointer to character after command name in cmdline
+static void set_cmd_addr_type(exarg_T *eap, char_u *p)
+{
+  // ea.addr_type for user commands is set by find_ucmd
+  if (!IS_USER_CMDIDX(eap->cmdidx)) {
+    if (eap->cmdidx != CMD_SIZE) {
+      eap->addr_type = cmdnames[(int)eap->cmdidx].cmd_addr_type;
+    } else {
+      eap->addr_type = ADDR_LINES;
+    }
+    // :wincmd range depends on the argument
+    if (eap->cmdidx == CMD_wincmd && p != NULL) {
+      get_wincmd_addr_type(skipwhite(p), eap);
+    }
+    // :.cc in quickfix window uses line number
+    if ((eap->cmdidx == CMD_cc || eap->cmdidx == CMD_ll) && bt_quickfix(curbuf)) {
+      eap->addr_type = ADDR_OTHER;
+    }
+  }
+}
+
+/// Set default command range based on the addr type of the command
+static void set_cmd_default_range(exarg_T *eap)
+{
+  buf_T *buf;
+
+  eap->line1 = 1;
+  switch (eap->addr_type) {
+  case ADDR_LINES:
+  case ADDR_OTHER:
+    eap->line2 = curbuf->b_ml.ml_line_count;
+    break;
+  case ADDR_LOADED_BUFFERS:
+    buf = firstbuf;
+    while (buf->b_next != NULL && buf->b_ml.ml_mfp == NULL) {
+      buf = buf->b_next;
+    }
+    eap->line1 = buf->b_fnum;
+    buf = lastbuf;
+    while (buf->b_prev != NULL && buf->b_ml.ml_mfp == NULL) {
+      buf = buf->b_prev;
+    }
+    eap->line2 = buf->b_fnum;
+    break;
+  case ADDR_BUFFERS:
+    eap->line1 = firstbuf->b_fnum;
+    eap->line2 = lastbuf->b_fnum;
+    break;
+  case ADDR_WINDOWS:
+    eap->line2 = LAST_WIN_NR;
+    break;
+  case ADDR_TABS:
+    eap->line2 = LAST_TAB_NR;
+    break;
+  case ADDR_TABS_RELATIVE:
+    eap->line2 = 1;
+    break;
+  case ADDR_ARGUMENTS:
+    if (ARGCOUNT == 0) {
+      eap->line1 = eap->line2 = 0;
+    } else {
+      eap->line2 = ARGCOUNT;
+    }
+    break;
+  case ADDR_QUICKFIX_VALID:
+    eap->line2 = (linenr_T)qf_get_valid_size(eap);
+    if (eap->line2 == 0) {
+      eap->line2 = 1;
+    }
+    break;
+  case ADDR_NONE:
+  case ADDR_UNSIGNED:
+  case ADDR_QUICKFIX:
+    iemsg(_("INTERNAL: Cannot use EX_DFLALL "
+            "with ADDR_NONE, ADDR_UNSIGNED or ADDR_QUICKFIX"));
+    break;
+  }
+}
+
+/// Check if command with ID handles '|' themselves
+bool command_handles_bar(cmdidx_T cmdidx)
```

Comment:
should be generated from `ex_cmds.lua`?

Reply:
It's a list of commands that don't have the `-bar` attribute and handle the `|` character by themselves, I found it in the command parsing code and moved it from there into its own function.

Reply:
On second thought maybe we shouldn't rely on this list and just check if the command doesn't have the `-bar` attribute? Though I'm not sure if that'll work perfectly since I'm not sure if this list contains all commands without the `-bar` attribute

Reply:
Okay so it seems that all commands that don't have the `-bar` attribute aren't in the list, and those that aren't in the list are commands that always treat `|` as their argument. So it depends on what meaning we want the `|` magic value to have? Do we want it to mean that `|` is treated as a command separator or do we want it to mean that `|` is treated as an argument

Reply:
> Do we want it to mean that `|` is treated as a command separator or do we want it to mean that `|` is treated as an argument.

If it's treated as an arg, that's "literal". If it's a separator that's "magic".

The purpose of `magic=[]` (i.e. _no magic_) is to say "I want all args to be passed literally to the command".

Reply:
> > Do we want it to mean that `|` is treated as a command separator or do we want it to mean that `|` is treated as an argument.
> 
> If it's treated as an arg, that's "literal". If it's a separator that's "magic".
> 
> The purpose of `magic=[]` (i.e. _no magic_) is to say "I want all args to be passed literally to the command". Commands can, of course, interpret the args any way they want.

In that case, I need to change the behavior a bit. But there's some commands where it's ambiguous whether `|` is treated literally or not, and it depends on the position of the `|`, so it's near impossible to get a fully accurate definitive list of which commands treat `|` specially. We could do the next best thing however and just put `|` as magic for all commands with the `-bar` attribute

Reply:
> We could do the next best thing however and just put | as magic for all commands with the -bar attribute

Sure let's start with that. Doesn't need to be perfect. This is already going to be super valuable and we can make it more granular later.

Reply:
Alright, done!